File: tools/server_observer/src/scheduler.cpp
#include "scheduler.hpp"
#include <iostream>
#include <algorithm>
#include <boost/asio/co_spawn.hpp>
#include <boost/asio/detached.hpp>

Scheduler::Scheduler(int max_parallel_updates,
                     int max_parallel_first_updates,
                     double update_interval,
                     int num_worker_threads)
    : max_parallel_updates_(max_parallel_updates)
    , max_parallel_first_updates_(max_parallel_first_updates)
    , update_interval_(std::chrono::milliseconds(static_cast<int64_t>(update_interval * 1000)))
    , num_worker_threads_(num_worker_threads)
    , active_coroutines_(0)
    , stop_flag_(false)
{
    // Initialize worker threads for update coroutines
    work_guard_ = std::make_unique<asio::executor_work_guard<asio::io_context::executor_type>>(
        asio::make_work_guard(io_context_)
    );

    std::cout << "Starting " << num_worker_threads_ << " scheduler worker threads" << std::endl;
    for (int i = 0; i < num_worker_threads_; ++i) {
        worker_threads_.emplace_back([this]() {
            io_context_.run();
        });
    }
}

Scheduler::~Scheduler() {
    if (!stop_flag_) {
        stop();
    }
}

void Scheduler::stop(int timeout_seconds) {
    std::cout << "Stopping Scheduler..." << std::endl;
    stop_flag_ = true;

    // Wait for all active coroutines to complete with a timeout
    std::cout << "Waiting for " << active_coroutines_.load()
              << " active coroutines to complete..." << std::endl;

    auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(timeout_seconds);
    while (active_coroutines_.load() > 0 &&
           std::chrono::steady_clock::now() < deadline) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    if (active_coroutines_.load() > 0) {
        std::cerr << "Warning: " << active_coroutines_.load()
                  << " coroutines still active after timeout" << std::endl;
    }

    // Stop the io_context and join worker threads
    work_guard_.reset();  // Release the work guard to allow io_context to finish
    io_context_.stop();   // Stop the io_context

    std::cout << "Waiting for " << worker_threads_.size()
              << " scheduler worker threads to finish..." << std::endl;

    // Join all worker threads
    for (auto& thread : worker_threads_) {
        if (thread.joinable()) {
            thread.join();
        }
    }

    std::cout << "Scheduler stopped successfully" << std::endl;
}

void Scheduler::schedule_update(ObservationSession* session) {
    // Thread-safe - can be called from worker threads (via handle_successful_update)
    std::lock_guard<std::mutex> lock(update_queue_lock_);
    update_queue_.insert({session->next_update_at, session});
}

void Scheduler::schedule_immediate_update(ObservationSession* session) {
    session->next_update_at = std::chrono::system_clock::now();
    schedule_update(session);
}

void Scheduler::schedule_next_update(ObservationSession* session, bool missed_update) {
    auto update_interval_ms = std::chrono::duration_cast<std::chrono::milliseconds>(update_interval_);
    int64_t offset_ms = calculate_offset_ms(session->game_id);
    
    auto now = std::chrono::system_clock::now();
    
    if (missed_update) {
        // Recalculate k based on current time
        auto time_since_epoch_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
            now.time_since_epoch()
        ).count();
        
        session->update_sequence_number = calculate_next_k(time_since_epoch_ms, offset_ms);
    } else {
        // Normal case: increment k by 1
        session->update_sequence_number++;
    }
    
    // Calculate next update time: k * update_interval + offset
    auto time_since_epoch = std::chrono::milliseconds(
        session->update_sequence_number * update_interval_ms.count() + offset_ms
    );
    session->next_update_at = std::chrono::system_clock::time_point(time_since_epoch);
    
    schedule_update(session);
}

int64_t Scheduler::calculate_offset_ms(int game_id) const {
    auto update_interval_ms = std::chrono::duration_cast<std::chrono::milliseconds>(update_interval_);
    return game_id % update_interval_ms.count();
}

int64_t Scheduler::calculate_next_k(int64_t current_time_ms, int64_t offset_ms) const {
    auto update_interval_ms = std::chrono::duration_cast<std::chrono::milliseconds>(update_interval_);
    int64_t interval_count = update_interval_ms.count();
    
    // Safety check: interval must be positive
    if (interval_count <= 0) {
        std::cerr << "Error: update_interval must be positive" << std::endl;
        return 1;  // Return safe default
    }
    
    // Handle case where current time is before the offset
    if (current_time_ms < offset_ms) {
        return 0;  // Next update is at k=0 (at offset_ms)
    }
    
    // Find smallest k where k * interval_count + offset_ms > current_time_ms
    // This is equivalent to: k > (current_time_ms - offset_ms) / interval_count
    // So k = floor((current_time_ms - offset_ms) / interval_count) + 1
    int64_t k = (current_time_ms - offset_ms) / interval_count + 1;
    
    return k;
}

void Scheduler::mark_first_update(int game_id) {
    // No lock needed - called only from main thread
    first_update_sessions_.insert(game_id);
}

void Scheduler::cleanup_first_update_tracking(int game_id) {
    // No lock needed - called only from main thread
    first_update_sessions_.erase(game_id);
    running_first_updates_.erase(game_id);
}

bool Scheduler::can_start_update(ObservationSession* session) {
    // No lock needed - called only from main thread
    if (first_update_sessions_.count(session->game_id)) {
        // Count currently running first updates
        int running_first_updates_count = static_cast<int>(running_first_updates_.size());

        if (running_first_updates_count >= max_parallel_first_updates_) {
            std::cout << "Deferring first update for game " << session->game_id
                     << " due to max parallel first updates limit (currently running: "
                     << running_first_updates_count << ")." << std::endl;
            return false;
        }
    }

    return true;
}

void Scheduler::mark_first_update_running(int game_id) {
    running_first_updates_.insert(game_id);
}

std::vector<ObservationSession*> Scheduler::get_due_updates() {
    auto now = std::chrono::system_clock::now();
    std::vector<ObservationSession*> ready;

    std::lock_guard<std::mutex> lock(update_queue_lock_);

    // Process sessions in time order from the multimap
    while (!update_queue_.empty()) {
        // Check if we've reached the limit of concurrent coroutines
        if (active_coroutines_.load() >= max_parallel_updates_) {
            break;
        }

        // Get the earliest scheduled session
        auto it = update_queue_.begin();
        auto* observer = it->second;

        // Check if it's due yet
        if (!observer->needs_update(now)) {
            // No more sessions are due (since they're time-ordered)
            break;
        }

        // Remove from queue
        update_queue_.erase(it);

        // Check if this update can start (respecting first-update limits)
        if (!can_start_update(observer)) {
            // Re-insert at the end with same time (will be retried next iteration)
            update_queue_.insert({observer->next_update_at, observer});
            continue;
        }

        // Mark as running first update if needed
        if (first_update_sessions_.count(observer->game_id)) {
            mark_first_update_running(observer->game_id);
        }

        ready.push_back(observer);
    }


    return ready;
}

void Scheduler::process_due_updates() {
    auto now = std::chrono::system_clock::now();

    std::chrono::system_clock::time_point next_due_time;
    bool queue_empty;

    {
        std::lock_guard<std::mutex> lock(update_queue_lock_);
        queue_empty = update_queue_.empty();
        if (!queue_empty) {
            next_due_time = update_queue_.begin()->first;
        }
    }

    // Check if queue is empty
    if (queue_empty) {
        // No updates pending, sleep briefly to avoid busy-waiting
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        return;
    }

    // Check if we're at the concurrency limit
    if (active_coroutines_.load() >= max_parallel_updates_) {
        // At limit, wait a short time for some to complete
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        return;
    }


    // Calculate wait time
    if (next_due_time > now) {
        auto wait_duration = next_due_time - now;
        double wait_seconds = std::chrono::duration<double>(wait_duration).count();

        // Cap the wait time at update_interval to allow periodic checks
        double update_interval_seconds = std::chrono::duration<double>(update_interval_).count();
        double actual_wait = std::min(wait_seconds, update_interval_seconds);

        if (actual_wait > 0.1) {  // Only log if waiting more than 100ms
            std::cout << "Waiting " << actual_wait
                     << " seconds for next due update..." << std::endl;
        }

        std::this_thread::sleep_for(std::chrono::duration<double>(actual_wait));
    }
}

void Scheduler::spawn_update_coroutine(asio::awaitable<void> update_coro) {
    asio::co_spawn(
        io_context_,
        std::move(update_coro),
        asio::detached
    );
}

void Scheduler::queue_new_session(int game_id, int scenario_id) {
    // Thread-safe - can be called from GameFinder thread
    std::lock_guard<std::mutex> lock(pending_sessions_lock_);
    pending_new_sessions_.push({game_id, scenario_id});
}

std::vector<std::pair<int, int>> Scheduler::get_pending_new_sessions() {
    // Called from main thread only
    std::vector<std::pair<int, int>> result;

    std::lock_guard<std::mutex> lock(pending_sessions_lock_);
    while (!pending_new_sessions_.empty()) {
        result.push_back(pending_new_sessions_.front());
        pending_new_sessions_.pop();
    }

    return result;
}

double Scheduler::get_seconds_until_next_due() const {
    std::lock_guard<std::mutex> lock(update_queue_lock_);

    if (update_queue_.empty()) {
        return 0.0;
    }

    auto now = std::chrono::system_clock::now();
    auto next_due = update_queue_.begin()->first;

    if (next_due <= now) {
        return 0.0;
    }

    auto duration = next_due - now;
    return std::chrono::duration<double>(duration).count();
}

void Scheduler::set_update_interval(double interval_seconds) {
    if (interval_seconds > 0.0) {
        update_interval_ = std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::duration<double>(interval_seconds));
        std::cout << "Scheduler: Updated update_interval to " << interval_seconds << " seconds" << std::endl;
    } else {
        std::cerr << "Scheduler: Invalid update_interval " << interval_seconds << ", must be > 0" << std::endl;
    }
}

void Scheduler::set_max_parallel_updates(int max_updates) {
    if (max_updates >= 1) {
        max_parallel_updates_ = max_updates;
        std::cout << "Scheduler: Updated max_parallel_updates to " << max_updates << std::endl;
    } else {
        std::cerr << "Scheduler: Invalid max_parallel_updates " << max_updates << ", must be >= 1" << std::endl;
    }
}

void Scheduler::set_max_parallel_first_updates(int max_first_updates) {
    if (max_first_updates >= 1) {
        max_parallel_first_updates_ = max_first_updates;
        std::cout << "Scheduler: Updated max_parallel_first_updates to " << max_first_updates << std::endl;
    } else {
        std::cerr << "Scheduler: Invalid max_parallel_first_updates " << max_first_updates 
                 << ", must be >= 1" << std::endl;
    }
}

void Scheduler::initialize_session_schedule(ObservationSession* session) {
    auto update_interval_ms = std::chrono::duration_cast<std::chrono::milliseconds>(update_interval_);
    int64_t offset_ms = calculate_offset_ms(session->game_id);
    
    auto now = std::chrono::system_clock::now();
    auto time_since_epoch_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()
    ).count();
    
    // Calculate smallest k where k * update_interval + offset > current_time
    session->update_sequence_number = calculate_next_k(time_since_epoch_ms, offset_ms);
    
    // Calculate next update time: k * update_interval + offset
    auto time_since_epoch = std::chrono::milliseconds(
        session->update_sequence_number * update_interval_ms.count() + offset_ms
    );
    session->next_update_at = std::chrono::system_clock::time_point(time_since_epoch);
}

File: services/server_observer/src/scheduler.rs
use crate::observation_session::ObservationSession;
use std::collections::{BTreeMap, HashSet, VecDeque};
use std::sync::atomic::{AtomicBool, AtomicI32, Ordering};
use std::sync::Mutex;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::time::sleep;

pub struct Scheduler {
    max_parallel_updates: AtomicI32,
    max_parallel_first_updates: AtomicI32,
    update_interval: Mutex<Duration>,
    update_queue: Mutex<BTreeMap<SystemTime, VecDeque<i32>>>,
    pending_priority_sessions: Mutex<VecDeque<(i32, i32)>>,
    pending_normal_sessions: Mutex<VecDeque<(i32, i32)>>,
    queued_priority: Mutex<HashSet<i32>>,
    queued_normal: Mutex<HashSet<i32>>,
    first_update_sessions: Mutex<HashSet<i32>>,
    running_first_updates: Mutex<HashSet<i32>>,
    active_coroutines: AtomicI32,
    stop_flag: AtomicBool,
}

impl Scheduler {
    pub fn new(
        max_parallel_updates: i32,
        max_parallel_first_updates: i32,
        update_interval: f64,
    ) -> Self {
        Self {
            max_parallel_updates: AtomicI32::new(max_parallel_updates.max(1)),
            max_parallel_first_updates: AtomicI32::new(max_parallel_first_updates.max(1)),
            update_interval: Mutex::new(Duration::from_secs_f64(update_interval.max(0.1))),
            update_queue: Mutex::new(BTreeMap::new()),
            pending_priority_sessions: Mutex::new(VecDeque::new()),
            pending_normal_sessions: Mutex::new(VecDeque::new()),
            queued_priority: Mutex::new(HashSet::new()),
            queued_normal: Mutex::new(HashSet::new()),
            first_update_sessions: Mutex::new(HashSet::new()),
            running_first_updates: Mutex::new(HashSet::new()),
            active_coroutines: AtomicI32::new(0),
            stop_flag: AtomicBool::new(false),
        }
    }

    pub fn stop(&self) {
        self.stop_flag.store(true, Ordering::SeqCst);
    }

    pub fn schedule_update(&self, session: &ObservationSession) {
        let mut queue = self.update_queue.lock().expect("scheduler queue mutex poisoned");
        queue
            .entry(session.next_update_at)
            .or_default()
            .push_back(session.game_id);
    }

    pub fn schedule_immediate_update(&self, session: &mut ObservationSession) {
        session.next_update_at = SystemTime::now();
        self.schedule_update(session);
    }

    pub fn schedule_next_update(&self, session: &mut ObservationSession, missed_update: bool) {
        let interval_ms = self.update_interval_ms();
        let offset_ms = self.calculate_offset_ms(session.game_id);
        let now = current_time_ms();

        if missed_update {
            session.update_sequence_number = self.calculate_next_k(now, offset_ms);
        } else {
            session.update_sequence_number += 1;
        }

        let due_ms = session.update_sequence_number * interval_ms + offset_ms;
        session.next_update_at = time_from_millis(due_ms);
        self.schedule_update(session);
    }

    pub fn initialize_session_schedule(&self, session: &mut ObservationSession) {
        let interval_ms = self.update_interval_ms();
        let offset_ms = self.calculate_offset_ms(session.game_id);
        let now_ms = current_time_ms();
        session.update_sequence_number = self.calculate_next_k(now_ms, offset_ms);
        let due_ms = session.update_sequence_number * interval_ms + offset_ms;
        session.next_update_at = time_from_millis(due_ms);
    }

    pub fn mark_first_update(&self, game_id: i32) {
        self.first_update_sessions
            .lock()
            .expect("scheduler first-update mutex poisoned")
            .insert(game_id);
    }

    pub fn cleanup_first_update_tracking(&self, game_id: i32) {
        self.first_update_sessions
            .lock()
            .expect("scheduler first-update mutex poisoned")
            .remove(&game_id);
        self.running_first_updates
            .lock()
            .expect("scheduler running-first-update mutex poisoned")
            .remove(&game_id);
    }

    pub fn queue_new_session(&self, game_id: i32, scenario_id: i32, is_priority: bool) {
        if is_priority {
            // If already queued priority, nothing to do.
            if self
                .queued_priority
                .lock()
                .expect("scheduler queued-priority mutex poisoned")
                .contains(&game_id)
            {
                return;
            }

            // If it was queued as normal, promote: remove from normal queue + sets.
            {
                let mut queued_normal = self
                    .queued_normal
                    .lock()
                    .expect("scheduler queued-normal mutex poisoned");
                if queued_normal.remove(&game_id) {
                    let mut normal_q = self
                        .pending_normal_sessions
                        .lock()
                        .expect("scheduler pending-normal mutex poisoned");
                    let mut new_q = VecDeque::with_capacity(normal_q.len());
                    while let Some(item) = normal_q.pop_front() {
                        if item.0 != game_id {
                            new_q.push_back(item);
                        }
                    }
                    *normal_q = new_q;
                }
            }

            self.pending_priority_sessions
                .lock()
                .expect("scheduler pending-priority mutex poisoned")
                .push_back((game_id, scenario_id));
            self.queued_priority
                .lock()
                .expect("scheduler queued-priority mutex poisoned")
                .insert(game_id);
        } else {
            // Don't queue if already queued (either queue).
            if self
                .queued_priority
                .lock()
                .expect("scheduler queued-priority mutex poisoned")
                .contains(&game_id)
            {
                return;
            }
            if self
                .queued_normal
                .lock()
                .expect("scheduler queued-normal mutex poisoned")
                .contains(&game_id)
            {
                return;
            }

            self.pending_normal_sessions
                .lock()
                .expect("scheduler pending-normal mutex poisoned")
                .push_back((game_id, scenario_id));
            self.queued_normal
                .lock()
                .expect("scheduler queued-normal mutex poisoned")
                .insert(game_id);
        }
    }

    pub fn get_pending_new_sessions_with_limits(
        &self,
        max_total: usize,
        max_normal: usize,
    ) -> Vec<(i32, i32, bool)> {
        if max_total == 0 {
            return Vec::new();
        }

        let mut out = Vec::new();
        out.reserve(max_total);

        {
            let mut q = self
                .pending_priority_sessions
                .lock()
                .expect("scheduler pending-priority mutex poisoned");
            while out.len() < max_total {
                let Some((game_id, scenario_id)) = q.pop_front() else {
                    break;
                };
                self.queued_priority
                    .lock()
                    .expect("scheduler queued-priority mutex poisoned")
                    .remove(&game_id);
                out.push((game_id, scenario_id, true));
            }
        }

        if out.len() >= max_total || max_normal == 0 {
            return out;
        }

        let mut normals_started = 0usize;
        let mut q = self
            .pending_normal_sessions
            .lock()
            .expect("scheduler pending-normal mutex poisoned");
        while out.len() < max_total && normals_started < max_normal {
            let Some((game_id, scenario_id)) = q.pop_front() else {
                break;
            };
            self.queued_normal
                .lock()
                .expect("scheduler queued-normal mutex poisoned")
                .remove(&game_id);
            out.push((game_id, scenario_id, false));
            normals_started += 1;
        }

        out
    }

    pub fn get_due_updates(&self) -> Vec<i32> {
        let now = SystemTime::now();
        let max_parallel = self.max_parallel_updates.load(Ordering::SeqCst);
        let mut ready = Vec::new();
        let mut queue = self.update_queue.lock().expect("scheduler queue mutex poisoned");

        loop {
            let in_flight = self.active_coroutines.load(Ordering::SeqCst) + ready.len() as i32;
            if in_flight >= max_parallel {
                break;
            }

            let Some((&next_due, _)) = queue.iter().next() else {
                break;
            };
            if next_due > now {
                break;
            }

            let game_id = pop_earliest(&mut queue).expect("earliest queue entry must exist");
            if !self.can_start_update(game_id) {
                queue.entry(next_due).or_default().push_back(game_id);
                break;
            }

            if self
                .first_update_sessions
                .lock()
                .expect("scheduler first-update mutex poisoned")
                .contains(&game_id)
            {
                self.running_first_updates
                    .lock()
                    .expect("scheduler running-first-update mutex poisoned")
                    .insert(game_id);
            }

            ready.push(game_id);
        }

        ready
    }

    pub async fn process_due_updates(&self) {
        if self.stop_flag.load(Ordering::SeqCst) {
            return;
        }

        let queue_state = {
            let queue = self.update_queue.lock().expect("scheduler queue mutex poisoned");
            queue.iter().next().map(|(k, _)| *k)
        };

        if queue_state.is_none() {
            sleep(Duration::from_millis(100)).await;
            return;
        }

        if self.active_coroutines.load(Ordering::SeqCst)
            >= self.max_parallel_updates.load(Ordering::SeqCst)
        {
            sleep(Duration::from_millis(100)).await;
            return;
        }

        let next_due = queue_state.expect("checked Some above");
        let now = SystemTime::now();
        if next_due > now {
            let wait = next_due.duration_since(now).unwrap_or(Duration::from_millis(0));
            let cap = *self
                .update_interval
                .lock()
                .expect("scheduler interval mutex poisoned");
            let actual_wait = wait.min(cap);
            if actual_wait > Duration::from_millis(0) {
                sleep(actual_wait).await;
            }
        }
    }

    pub fn increment_active_coroutines(&self) {
        self.active_coroutines.fetch_add(1, Ordering::SeqCst);
    }

    pub fn decrement_active_coroutines(&self) {
        self.active_coroutines.fetch_sub(1, Ordering::SeqCst);
    }

    pub fn set_update_interval(&self, interval_seconds: f64) {
        if interval_seconds > 0.0 {
            *self
                .update_interval
                .lock()
                .expect("scheduler interval mutex poisoned") =
                Duration::from_secs_f64(interval_seconds);
        }
    }

    pub fn set_max_parallel_updates(&self, max_updates: i32) {
        if max_updates >= 1 {
            self.max_parallel_updates
                .store(max_updates, Ordering::SeqCst);
        }
    }

    pub fn set_max_parallel_first_updates(&self, max_first_updates: i32) {
        if max_first_updates >= 1 {
            self.max_parallel_first_updates
                .store(max_first_updates, Ordering::SeqCst);
        }
    }

    fn update_interval_ms(&self) -> i64 {
        self.update_interval
            .lock()
            .expect("scheduler interval mutex poisoned")
            .as_millis()
            .max(1) as i64
    }

    fn calculate_offset_ms(&self, game_id: i32) -> i64 {
        game_id as i64 % self.update_interval_ms()
    }

    fn calculate_next_k(&self, current_time_ms: i64, offset_ms: i64) -> i64 {
        let interval = self.update_interval_ms();
        if interval <= 0 {
            return 1;
        }
        if current_time_ms < offset_ms {
            return 0;
        }
        (current_time_ms - offset_ms) / interval + 1
    }

    fn can_start_update(&self, game_id: i32) -> bool {
        let first = self
            .first_update_sessions
            .lock()
            .expect("scheduler first-update mutex poisoned");
        if !first.contains(&game_id) {
            return true;
        }
        drop(first);

        let running = self
            .running_first_updates
            .lock()
            .expect("scheduler running-first-update mutex poisoned");
        let max_first = self.max_parallel_first_updates.load(Ordering::SeqCst);
        (running.len() as i32) < max_first
    }
}

fn pop_earliest(queue: &mut BTreeMap<SystemTime, VecDeque<i32>>) -> Option<i32> {
    let key = *queue.keys().next()?;
    let ids = queue.get_mut(&key)?;
    let id = ids.pop_front();
    if ids.is_empty() {
        queue.remove(&key);
    }
    id
}

fn current_time_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or_default()
}

fn time_from_millis(ms: i64) -> SystemTime {
    if ms <= 0 {
        return UNIX_EPOCH;
    }
    UNIX_EPOCH + Duration::from_millis(ms as u64)
}
